fabric/operations.py

```diff
@@ -244,6 +244,34 @@ def prompt(text, key=None, default='', validate=None):
     # And return the value, too, just in case someone finds that useful.
     return value
 
+def _remote_tilde_expand (remote_path=None, use_sudo=False):
+    """
+    Performs tilde (~) expansion on a remote host
+
+    Theoretically you could just replace any occurrence ~/ with
+    /home/{env.user} and ~username/ with /home/{username}. However,
+    this works on the presumption that all home directories are
+    stored under /home. Currently this pulls the remote home directory
+    directly from /etc/passwd by executing awk, but there may be a
+    more elegant way of doing this.
+    """
+    if remote_path.startswith('~'):
+        if not '/' in remote_path:
+            expand_user = remote_path[1:]
+            remote_path = ''
+        else:
+            expand_user = remote_path[1:remote_path.index('/')]
+            remote_path = remote_path[remote_path.index('/')+1:]
+
+        if expand_user == '':
+            expand_user = env.sudo_user if use_sudo else env.user;
+
+        expanded = _run_command (command="awk -F: '$1 == \"%s\" {print $6}' /etc/passwd" % (expand_user), sudo=use_sudo, quiet=True)
```

Comment:
I don't think you need sudo to read /etc/passwd, it is world-readable on all linux systems I've seen